### PR TITLE
Test/adding model tests

### PR DESCRIPTION
**Pull request by Dearop**
# Pull Request Request: Adding Model Tests and Test Modules

## Description

This PR introduces comprehensive unit tests and supporting mock modules to enhance the test coverage of the repositories and related components within the application. The focus is on ensuring the reliability and correctness of Firestore-based repository implementations using dependency injection and Hilt testing utilities.

### 1. Unit Tests for Firestore Repositories:
* Added tests for FoodItemRepositoryFirestore, HouseholdRepositoryFirestore, InvitationRepositoryFirestore, and UserRepositoryFirestore.
* Covered scenarios such as adding, updating, deleting, and retrieving entities.
* Tested error handling and edge cases like empty inputs and invalid documents.
### 2. Test Modules with Mock Dependencies:
* Created TestFirestoreModule, TestFoodItemModule, TestHouseHoldModule, TestInvitationModule, and TestUserModule for providing mocked dependencies during testing.
* These modules use Mockito to mock Firestore and FirebaseAuth interactions.
### 3. AndroidManifest Update for Hilt Testing:
* Modified the manifest to use HiltTestApplication for better compatibility with Hilt testing.
### 4. Improved Test Coverage:
* Added tests for Firestore listeners and caching mechanisms.
* Validated the state management and caching behavior of repositories under various conditions.

## Testing

* The tests were run using Robolectric and passed successfully.
* Verified integration with Hilt and ensured proper replacement of production modules with test modules.

### Repositories and Modules:
* Updated constructors for repositories to use @Inject annotations for easier dependency injection.
### Tests:
* Added new test classes for each repository.
* Mocked Firestore collections, documents, and listeners to simulate real-world behavior.

### CODE REVIEW

File: app/src/main/java/com/android/shelfLife/model/invitations/InvitationRepositoryFirestore.kt [hunk 1]
```diff
@@ -8,12 +8,15 @@ import com.google.firebase.firestore.DocumentSnapshot
...
 import kotlinx.coroutines.tasks.await
 
-open class InvitationRepositoryFirestore(
-    private val db: FirebaseFirestore,
-    private val auth: FirebaseAuth
-) : InvitationRepository {
+@Singleton
+open class InvitationRepositoryFirestore
+@Inject
+constructor(private val db: FirebaseFirestore, private val auth: FirebaseAuth) :
```
Comment (frostymate) [inline app/src/main/java/com/android/shelfLife/model/invitations/InvitationRepositoryFirestore.kt:18]: I see most changes are about injecting constructors into the base of the classes, is this supposed to be helpful with the testing>

File: app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+import org.mockito.Mockito.*
+import org.robolectric.RobolectricTestRunner
+
+@HiltAndroidTest
+@RunWith(RobolectricTestRunner::class)
+class FoodItemRepositoryFirestoreTest {
+
+  @get:Rule var hiltRule = HiltAndroidRule(this)
+
+  @Inject lateinit var foodItemRepository: FoodItemRepositoryFirestore
```
Comment (frostymate) [inline app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt:31]: i see we are not injecting some repositories like FoodItemRepositoryFirestore

File: app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+  @Before
+  fun setUp() {
+    hiltRule.inject()
+    val mockContext = mock(Context::class.java)
+    `when`(mockContext.getApplicationContext()).thenReturn(mock(Context::class.java))
+    mockCollection = mock(CollectionReference::class.java)
+    mockDocument = mock(DocumentReference::class.java)
+
+    // Mock Firestore collection and document calls
+    `when`(mockFirestore.collection(anyString())).thenReturn(mockCollection)
```
Comment (frostymate) [inline app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt:47]: This is a good way of making sure calls to the firebase are stored properly

File: app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    val updatedItem = originalItem.copy(location = FoodStorageLocation.FREEZER)
+
+    `when`(mockCollection.document("item321")).thenReturn(mockDocument)
+    `when`(mockDocument.set(updatedItem)).thenReturn(Tasks.forResult(null))
+
+    // Act
+    foodItemRepository.updateFoodItem("household1", updatedItem)
+
+    // Assert
+    val cachedItems = foodItemRepository.foodItems.value
```
Comment (frostymate) [inline app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt:187]: The introduction to test the caching is a good feature

File: app/src/test/java/com/android/shelflife/model/FoodItemsTest.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    // Assert
+    assertNull(foodItem)
+  }
+
+  @Test
+  fun `test DocumentSnapshot toFoodItem conversion with invalid data`() {
+    // Arrange
+    val mockSnapshot = mock(DocumentSnapshot::class.java)
+    `when`(mockSnapshot.getString("uid")).thenReturn(null) // Missing UID
```
Comment (frostymate) [inline app/src/test/java/com/android/shelflife/model/FoodItemsTest.kt:84]: Good testing for error handling and for if the uid is missing here

**Review by frostymate — APPROVED** [review]
### Code Review for Pull Request #299

#### General Overview
This pull request introduces dependency injection using the `@Inject` annotation across several repository classes. It also includes tests for these changes and mocks for Firestore and Firebase Authentication. This are the new tests that we are going to base the whole app for. When we previously redid the whole architecture we did not do test due to the urgency of making sure our new architecture was usable to add the missing feature we had. Now that it is up and running we are focusing on making tests as those changed as well in order to improve our line converage percentage. This is a great start on doing that. I would say that just we careful with the injections and the way that we now "mock" stuff, it seems quite complex so some documentation could help me and other colleagues navigate throught the elements. However, it seems to pass the CI and when I was reviewing the tests I found them to be quite through. Good Job Paul!

#### Detailed Review
2. **FoodItemRepositoryFirestore.kt**
    - **Change**: The class now uses `@Inject` for the constructor and changes the visibility of `convertToFoodItem` method to `internal`.
    - **Comment**: The use of `@Inject` is a good move for dependency injection, enhancing testability. Changing `convertToFoodItem` to `internal` is a sensible decision for encapsulation.

3. **HouseHoldRepositoryFirestore.kt**
    - **Change**: Added `@Inject` to the constructor.
    - **Comment**: This follows the same pattern as the `FoodItemRepositoryFirestore` class, improving consistency and testability.

5. **Test Modules**
    - **Changes**: Added or modified test modules to use `@TestInstallIn` replacing original modules with mocked versions.
    - **Comment**: These changes are excellent for testing, ensuring that tests are isolated and do not rely on actual Firestore or Firebase Authentication services.

6. **Test Classes**
    - **Changes**: New test classes for different repositories.
    - **Comment**: The tests are comprehensive, covering various scenarios including error handling and CRUD operations. This significantly improves the codebase's reliability and maintainability.

File: app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt (hunk 2 above)
Comment (frostymate) [reply app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt:31]: *are

File: app/src/main/java/com/android/shelfLife/model/invitations/InvitationRepositoryFirestore.kt (hunk 1 above)
Comment (Dearop) [reply app/src/main/java/com/android/shelfLife/model/invitations/InvitationRepositoryFirestore.kt:18]: With these changes we are able to inject the dependencies of those models into the Tests

File: app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt (hunk 2 above)
Comment (Dearop) [reply app/src/test/java/com/android/shelflife/model/FoodItemRepositoryFirestoreTest.kt:31]: Yes to use HILT
